kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,436 @@
+				"ON CONFLICT (key) WHERE parent_id IS NULL "+
+				"DO UPDATE SET value=$2 "+
+				"WHERE "+b.table+".value IS NOT NULL",
+			key, value,
+		)
+	} else {
+		// ON CONFLICT requires the WHERE parent_id NOT IS NULL hint to
+		// let Postgres find the non-NULL-parent_id unique index
+		// (<table>_up).
+		result, err = b.tx.Exec(
+			"INSERT INTO "+b.table+" (key, value, parent_id) "+
+				"VALUES($1, $2, $3) "+
+				"ON CONFLICT (key, parent_id) "+
+				"WHERE parent_id IS NOT NULL "+
+				"DO UPDATE SET value=$2 "+
+				"WHERE "+b.table+".value IS NOT NULL",
+			key, value, b.id,
+		)
+	}
+	if err != nil {
+		return err
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rows != 1 {
+		return walletdb.ErrIncompatibleValue
+	}
+
+	return nil
+}
+
+// Delete deletes the key/value pointed to by the passed key.
+// Returns ErrKeyRequired if the passed key is empty.
+func (b *readWriteBucket) Delete(key []byte) error {
+	if key == nil {
+		return nil
+	}
+	if len(key) == 0 {
+		return walletdb.ErrKeyRequired
+	}
+
+	// Check to see if a bucket with this key exists.
+	var dummy int
+	err := b.tx.QueryRow(
+		"SELECT 1 FROM "+b.table+" WHERE "+parentSelector(b.id)+
+			" AND key=$1 AND value IS NULL", key,
+	).Scan(&dummy)
+	switch {
+	// No bucket exists, proceed to deletion of the key.
+	case err == sql.ErrNoRows:
+
+	case err != nil:
+		return err
+
+	// Bucket exists.
+	default:
+		return walletdb.ErrIncompatibleValue
+	}
+
+	_, err = b.tx.Exec(
+		"DELETE FROM "+b.table+" WHERE key=$1 AND "+
+			parentSelector(b.id)+" AND value IS NOT NULL",
+		key,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ReadWriteCursor returns a new read-write cursor for this bucket.
+func (b *readWriteBucket) ReadWriteCursor() walletdb.ReadWriteCursor {
+	return newReadWriteCursor(b)
+}
+
+// Tx returns the buckets transaction.
+func (b *readWriteBucket) Tx() walletdb.ReadWriteTx {
+	return b.tx
+}
+
+// NextSequence returns an autoincrementing sequence number for this bucket.
+// Note that this is not a thread safe function and as such it must not be used
+// for synchronization.
+func (b *readWriteBucket) NextSequence() (uint64, error) {
+	seq := b.Sequence() + 1
+
+	return seq, b.SetSequence(seq)
+}
+
+// SetSequence updates the sequence number for the bucket.
+func (b *readWriteBucket) SetSequence(v uint64) error {
+	var (
+		query     string
+		queryArgs []interface{}
+	)
+
+	if b.id == nil {
+		// There is no parent row to point to, so set the sequence
+		// number in the system table.
+		query = "INSERT INTO " + b.tx.db.sequenceTableName +
+			" (table_name, sequence) VALUES($1, $2) " +
+			"ON CONFLICT (table_name) DO UPDATE SET sequence=$2"
+		queryArgs = []interface{}{b.table, int64(v)}
+	} else {
+		// Update the sequence field on the parent row.
+		query = "UPDATE " + b.table + " SET sequence=$2 WHERE id=$1"
+		queryArgs = []interface{}{b.id, int64(v)}
+	}
+
+	result, err := b.tx.Exec(
```

Comment:
Any reason we use the arg query in some instances (like here) and in others we do in-line string formatting?

Reply:
Yes, because the query substitution only works for some parts of the query iirc. So `where key=$1` works, but `where $1` doesn't.